plugins/analytics/public/shodan_api.py

```diff
@@ -0,0 +1,67 @@
+from pprint import pformat
+import shodan
+from core.analytics import OneShotAnalytics
+from core.entities import Company
+from core.observables import Text, Hostname, Tag
+
+
+class ShodanApi(object):
+    settings = {
+        "shodan_api_key": {
+            "name": "Shodan API Key",
+            "description": "API Key provided by Shodan.io."
+        }
+    }
+
+    @staticmethod
+    def fetch(observable, api_key):
+        try:
+            r = shodan.Shodan(api_key).host(observable.value)
+        except shodan.APIError, e:
+            print 'Error: %s' % e
+        return r
+
+
+class ShodanQuery(OneShotAnalytics, ShodanApi):
+    default_values = {
+        "name": "Shodan",
+        "description": "Perform a Shodan query on the IP address and tries to"
+                       " extract relevant information."
+    }
+
+    ACTS_ON = "Ip"
+
+    @staticmethod
+    def analyze(ip, results):
+        links = set()
+        result = ShodanApi.fetch(ip, results.settings['shodan_api_key'])
+        results.update(raw=pformat(result))
+
+        if 'tag' in result and result['tag'] is not None:
+            o_tag = Tag.get_or_create(result['tag'])
```

Comment:
I do not know Shodan that well, can you give example of tags or observable with meaningful tags ?

You just have to do ```ip.tag(result['tag'])```. This will take care of creating the tag is it doesn't exist. You can pass this method a single string (tag value), or a list of strings.